README.md
# docker-cache-registry
# test

## Installation

Cert-manager is used to issue TLS certificate for the mutating webhook. It is thus required to install it first.

1. [Install](https://cert-manager.io/docs/installation/) cert-manager.
1. Install the helm chart `helm upgrade --install --namespace cache-registry-system cache-registry ./helm/cache-registry/ --values=./values.yaml`.

## Pod filtering

There are 2 ways to filter pods from which images should be cached.

- The first and most basic way is to add the label `cache-registry.enix.io/image-caching-policy: ignore` on pods that should be ignored.
- The second way is to define the value `controllers.webhook.objectSelector.matchExpressions` in helm `values.yaml` configuration file.

Those parameters are used by the `MutatingWebhookConfiguration` to filter pods that needs to be updated. Once images from those pods are rewritten, a label will be added to them so the Pod controller will create CachedImages custom resources. The CachedImages controller will then cache those images.

## Cache persistance & garbage collecting

Persistance is disabled by default. It requires a CSI plugin to be installed on the cluster to be enabled. It is then configured through the `values.yaml` helm release configuration file in `registry.persistance`.

When a CachedImage expires because it is not used anymore by the cluster, the image is deleted from the registry. But it only delete **reference files** like tags, not blobs that accounts for the most storage usage. [Garbage collection](https://docs.docker.com/registry/garbage-collection/) allows to remove those blobs and free space. The garbage collecting job can be configured to run thanks to the `registry.garbageCollectionSchedule` configuration in a cron-like format. It is disabled by default as running garbage collection without persistency configured would just empty the cache registry as described in the below section.

### ⚠️ Limitations

Garbage collection can only run when the registry is read-only or not running at all to prevent corrupted images as described in the [documentation](https://docs.docker.com/registry/garbage-collection/). Thus, when the garbage collection job runs, it first stops any running instance of the cache registry before doing garbage collection. During this period of time, all pulls are proxified to the upstream registry so operation can continue smoothly.

Be careful, running garbage collection while not having persistance configured would simply empty the cache registry since its pod is destroyed during the operation and it is thus not recommanded for production setups.


CHANGELOG.md
# [1.5.0](https://gitlab.enix.io/products/docker-cache-registry/compare/v1.4.2...v1.5.0) (2022-12-30)


### Bug Fixes

* **helm:** don't trigger garbage collection if persistence isn't enabled ([f16d5a1](https://gitlab.enix.io/products/docker-cache-registry/commit/f16d5a1daa715979e74e5a75c2a9b862dcecc0ea))
* **helm:** ensure registry directories before garbage collection ([049c7a1](https://gitlab.enix.io/products/docker-cache-registry/commit/049c7a1c243cf95225b11df49065c495ca7dec37))


### Features

* **controllers:** support multiple pull secrets when caching ([5acd51b](https://gitlab.enix.io/products/docker-cache-registry/commit/5acd51b5387e5420606e8693293300440acfd9fc))
* **proxy:** support multiple pull secrets when proxying ([bde0dd6](https://gitlab.enix.io/products/docker-cache-registry/commit/bde0dd65965f464e088ec1f9a232308d44a684a8)), closes [#25](https://gitlab.enix.io/products/docker-cache-registry/issues/25)

## [1.4.2](https://gitlab.enix.io/products/docker-cache-registry/compare/v1.4.1...v1.4.2) (2022-12-06)


### Bug Fixes

* **controllers:** fix pod count ([b1c0d86](https://gitlab.enix.io/products/docker-cache-registry/commit/b1c0d8610d8a23eda487d1ba713ae00d10013a66)), closes [#35](https://gitlab.enix.io/products/docker-cache-registry/issues/35)
* **helm:** Add default tolerations for proxy service ([a1b26b2](https://gitlab.enix.io/products/docker-cache-registry/commit/a1b26b264aba60ed9bded013e0bb30cf3eabea76))
* **helm:** Use correct controller labels for service selector ([4c7479a](https://gitlab.enix.io/products/docker-cache-registry/commit/4c7479af4b496132ddf29bdd0a78d3d4de0175db))

## [1.4.1](https://gitlab.enix.io/products/docker-cache-registry/compare/v1.4.0...v1.4.1) (2022-12-01)


### Bug Fixes

* **controllers:** add -latest to CachedImage names when tag is missing ([0b9b9b4](https://gitlab.enix.io/products/docker-cache-registry/commit/0b9b9b43c0625936c9df3de5451b22bc071f1078))
* **controllers:** hash repository labels longer than 63 chars ([d797e09](https://gitlab.enix.io/products/docker-cache-registry/commit/d797e091b3311a1bb8aa4cc6c60be2606a71924a)), closes [#33](https://gitlab.enix.io/products/docker-cache-registry/issues/33)
* **proxy:** allow variadic components in images names ([611f29d](https://gitlab.enix.io/products/docker-cache-registry/commit/611f29d3ec83599e5176e4f5a14545a4cb722a8f)), closes [#32](https://gitlab.enix.io/products/docker-cache-registry/issues/32)
* keep capital letters in sanitized names and make them lowercase ([212326b](https://gitlab.enix.io/products/docker-cache-registry/commit/212326b3d7945a094629600c6df4cdea3ce6790d))

# [1.4.0](https://gitlab.enix.io/products/docker-cache-registry/compare/v1.3.1...v1.4.0) (2022-11-24)


### Bug Fixes

* **controllers:** check for 404 when checking image existence ([db9b3ee](https://gitlab.enix.io/products/docker-cache-registry/commit/db9b3ee201964fcf0f05a88d32f01ad1bdb5e71c))
* **controllers:** don't trigger image caching again after deleting it ([dd90834](https://gitlab.enix.io/products/docker-cache-registry/commit/dd90834482d6db15166d4887cbde989375663fa2))
* **webhook:** prevent pod creation if webhook fails ([f2e3e02](https://gitlab.enix.io/products/docker-cache-registry/commit/f2e3e02eed55166447f31b36e37392061daa9a3a))


### Features

* **controllers:** CachedImage events ([e3b5e12](https://gitlab.enix.io/products/docker-cache-registry/commit/e3b5e127711cf3be4d4a27e98f074eb4cf57978f)), closes [#28](https://gitlab.enix.io/products/docker-cache-registry/issues/28)
* **crd:** add isCached in CachedImage additionalPrinterColumns ([032b527](https://gitlab.enix.io/products/docker-cache-registry/commit/032b527b52944a2dc6c0c824ea5f0df67426976e))
* **crd:** improve CachedImage status and columns ([f6d82a4](https://gitlab.enix.io/products/docker-cache-registry/commit/f6d82a4f278897a0b55b16dd34b0f5bcf46cfb52)), closes [#29](https://gitlab.enix.io/products/docker-cache-registry/issues/29)

## [1.3.1](https://gitlab.enix.io/products/docker-cache-registry/compare/v1.3.0...v1.3.1) (2022-10-19)


### Bug Fixes

* **helm:** support for k8s 1.20 ([2afa898](https://gitlab.enix.io/products/docker-cache-registry/commit/2afa89836468814231a6db4d59e1ba9581d561ea))
* **proxy:** handle authentication when proxying registry ([dec04ea](https://gitlab.enix.io/products/docker-cache-registry/commit/dec04ea9d57630b9852caf5fe214b114c5305dd5)), closes [#23](https://gitlab.enix.io/products/docker-cache-registry/issues/23)
* **proxy:** suppress errors on cancel pull ([73bb2a8](https://gitlab.enix.io/products/docker-cache-registry/commit/73bb2a8d620b9db27655cf87f022f87d23427581))

# [1.3.0](https://gitlab.enix.io/products/docker-cache-registry/compare/v1.2.1...v1.3.0) (2022-10-07)


### Bug Fixes

* **ci:** Helm jobs trigger ([9880881](https://gitlab.enix.io/products/docker-cache-registry/commit/9880881dbcd624d82734e9d9bed21e1d7a0414d3))
* **helm:** fix webhook object selector value usage ([70c12b5](https://gitlab.enix.io/products/docker-cache-registry/commit/70c12b5d5784f2ef0cf7fc420c80ab0b23968f20))
* **proxy:** fallback to origin registry when cache registry is down ([b438cd1](https://gitlab.enix.io/products/docker-cache-registry/commit/b438cd185ae7b0bbf66b0a13de7a5bb5488e7761))


### Features

* **cache:** garbage collection ([1b5dfd9](https://gitlab.enix.io/products/docker-cache-registry/commit/1b5dfd9095a9dff480aae73db458fa4e0f258061)), closes [#12](https://gitlab.enix.io/products/docker-cache-registry/issues/12)

## [1.2.1](https://gitlab.enix.io/products/docker-cache-registry/compare/v1.2.0...v1.2.1) (2022-09-27)


### Bug Fixes

* **controller:** filter out non-rewritten pods in controllers ([20e0552](https://gitlab.enix.io/products/docker-cache-registry/commit/20e0552ecab687766b129d95cd9f71eb41b9b4f3))
* **controller:** fix missing init containers in CachedImages + tests ([582ded8](https://gitlab.enix.io/products/docker-cache-registry/commit/582ded8c8b1d5ef868099d77abcf246b19f11928))

cmd/cache/main.go
/*
Copyright 2021.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"os"
	"time"

	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
	// to ensure that exec-entrypoint and run can make use of them.
	_ "k8s.io/client-go/plugin/pkg/client/auth"

	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/healthz"
	"sigs.k8s.io/controller-runtime/pkg/log/zap"
	"sigs.k8s.io/controller-runtime/pkg/webhook"

	kuikenixiov1 "github.com/enix/kube-image-keeper/api/v1"
	"github.com/enix/kube-image-keeper/controllers"
	"github.com/enix/kube-image-keeper/internal/scheme"
	//+kubebuilder:scaffold:imports
)

var (
	setupLog = ctrl.Log.WithName("setup")
)

func main() {
	var metricsAddr string
	var enableLeaderElection bool
	var probeAddr string
	var expiryDelay uint
	var proxyPort int
	var ignoreNamespace string
	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
	flag.BoolVar(&enableLeaderElection, "leader-elect", false,
		"Enable leader election for controller manager. "+
			"Enabling this will ensure there is only one active controller manager.")
	flag.UintVar(&expiryDelay, "expiry-delay", 30, "The delay in days before deleting an unused CachedImage.")
	flag.IntVar(&proxyPort, "proxy-port", 8082, "The port where the proxy is listening on this machine.")
	flag.StringVar(&ignoreNamespace, "ignore-namespace", "kuik-system", "The address the probe endpoint binds to.")
	opts := zap.Options{
		Development: true,
	}
	opts.BindFlags(flag.CommandLine)
	flag.Parse()

	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))

	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
		Scheme:                 scheme.NewScheme(),
		MetricsBindAddress:     metricsAddr,
		Port:                   9443,
		HealthProbeBindAddress: probeAddr,
		LeaderElection:         enableLeaderElection,
		LeaderElectionID:       "a046788b.kuik.enix.io",
	})
	if err != nil {
		setupLog.Error(err, "unable to start manager")
		os.Exit(1)
	}

	if err = (&controllers.CachedImageReconciler{
		Client:   mgr.GetClient(),
		Scheme:   mgr.GetScheme(),
		Recorder: mgr.GetEventRecorderFor("cachedimage-controller"),
	}).SetupWithManager(mgr); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "CachedImage")
		os.Exit(1)
	}
	if err = (&controllers.PodReconciler{
		Client:      mgr.GetClient(),
		Scheme:      mgr.GetScheme(),
		ExpiryDelay: time.Duration(expiryDelay*24) * time.Hour,
	}).SetupWithManager(mgr); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "Pod")
		os.Exit(1)
	}
	imageRewriter := kuikenixiov1.ImageRewriter{
		Client:          mgr.GetClient(),
		IgnoreNamespace: ignoreNamespace,
		ProxyPort:       proxyPort,
	}
	mgr.GetWebhookServer().Register("/mutate-core-v1-pod", &webhook.Admission{Handler: &imageRewriter})
	//+kubebuilder:scaffold:builder

	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
		setupLog.Error(err, "unable to set up health check")
		os.Exit(1)
	}
	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
		setupLog.Error(err, "unable to set up ready check")
		os.Exit(1)
	}

	setupLog.Info("starting manager")
	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
		setupLog.Error(err, "problem running manager")
		os.Exit(1)
	}
}
